diff --git a/tl/generate_tl.py b/tl/generate_tl.py
index eeb3d53..469a2c8 100644
--- a/tl/generate_tl.py
+++ b/tl/generate_tl.py
@@ -1524,6 +1524,10 @@ enum {\n\
 ' + factories + '\n\
 ' + ('} // namespace ' + globalNamespace + '\n' if globalNamespace != '' else '')
 
+  # --- AyuGram hook
+  header = header.replace('''	[[nodiscard]] bool is_noforwards() const;''', '''	[[nodiscard]] bool is_noforwards() const;\n	[[nodiscard]] bool is_ayuNoforwards() const;''')
+  header = header.replace('''	[[nodiscard]] bool is_restricted() const;''', '''	[[nodiscard]] bool is_restricted() const;\n	[[nodiscard]] bool is_ayuRestricted() const;''')
+
   source = '\
 // WARNING! All changes made in this file will be lost!\n\
 // Created from ' + inputNames + ' by \'generate.py\'\n\
@@ -1544,6 +1548,27 @@ public:\n\
 ' + methods + '\n\
 ' + ('} // namespace ' + globalNamespace + '\n' if globalNamespace != '' else '')
 
+  # --- AyuGram hook
+  rrr1 = re.compile(r'''bool MTPD(?P<class>.+?)::is_noforwards\(\) const {
+	return _flags\.v & Flag::f_noforwards;
+}''')
+  rrr2 = re.compile(r'''bool MTPD(?P<class>.+?)::is_restricted\(\) const {
+	return _flags\.v & Flag::f_restricted;
+}''')
+
+  source = rrr1.sub(r'''bool MTPD\g<class>::is_noforwards() const {
+	return false;
+}
+bool MTPD\g<class>::is_ayuNoforwards() const {
+	return _flags.v & Flag::f_noforwards;
+}''', source)
+  source = rrr2.sub(r'''bool MTPD\g<class>::is_restricted() const {
+	return false;
+}
+bool MTPD\g<class>::is_ayuRestricted() const {
+	return _flags.v & Flag::f_restricted;
+}''', source)
+
   conversionHeaderFrom = '\
 // WARNING! All changes made in this file will be lost!\n\
 // Created from ' + inputNames + ' by \'generate.py\'\n\
